### ISSUE TITLE
Add type hints

### ISSUE THREAD

**Issue by giswqs** [opengeos/leafmap#324]
Reference:

- https://github.com/python-visualization/folium/pull/1297/files

**Comment by krishnaglodha** [comment]
I can take up the task if needed

**Comment by giswqs** [comment]
Contributions are welcome!

**Comment by krishnaglodha** [comment]
@giswqs , can you please confirm if this is correct way to move forward ? So I can confidently work on all files

https://github.com/krishnaglodha/leafmap/blob/fix/addtyping/leafmap/basemaps.py

https://github.com/krishnaglodha/leafmap/blob/fix/addtyping/leafmap/bokehmap.py

**Comment by giswqs** [comment]
Looks good to me. One thing to keep in mind is to avoid importing external libraries that are not required dependencies (see requirements.txt) at the beginning of each module. 

Thank you for helping with this.

**Comment by krishnaglodha** [comment]
Great thank you for confirmation.

I’ll keep it in my mind about dependencies

Thank you,
Krishna G. Lodha
Personal email : ***@***.***
Office email : ***@***.***

**Comment by krishnaglodha** [comment]
I'm trying to run tests after setting up project as 
1. git clone 
2. python setup.py install
3. pytest . 

getting following error

```
E   Exception: No module named 'ipyleaflet'
-------------------------------------------- Captured stdout --------------------------------------------
Please restart Jupyter kernel after installation if you encounter any errors when importing leafmap.
======================================== short test summary info ========================================
ERROR tests/test_colormaps.py - Exception: No module named 'ipyleaflet'
ERROR tests/test_common.py
ERROR tests/test_foliumap.py - Exception: No module named 'ipyleaflet'
ERROR tests/test_leafmap.py - Exception: No module named 'ipyleaflet'
ERROR tests/test_osm.py
ERROR tests/test_toolbar.py - Exception: No module named 'ipyleaflet'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 6 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=========================================== 6 errors in 0.65s ===========================================
```

**Comment by giswqs** [comment]
Try this: 

```bash
pip install -e .
python -m unittest discover tests
```

You can also submit the pull request. GitHub Actions will take care of testing.

**Comment by krishnaglodha** [comment]
Hi @giswqs , I can see param having default value as `True` but in comment it says `str` 
https://github.com/opengeos/leafmap/blob/master/leafmap/foliumap.py#L286

**Comment by giswqs** [comment]
Good catch. They are typos. Should be bool, not str.

**Comment by giswqs** [comment]
We can't drop support for 3.8 yet. Skip checking the function if needed

**Comment by krishnaglodha** [comment]
so should I remove type check for `Tuple` all together  ?

**Comment by giswqs** [comment]
Is it related to this issue?

https://github.com/python/typing/issues/1230

**Comment by krishnaglodha** [comment]
Hi @giswqs , can you please confirm if MAPBOX_APIKEY is available in environment for CI file, I'm getting following error https://github.com/opengeos/leafmap/actions/runs/5597313369/jobs/10235495642#step:10:623

**Comment by giswqs** [comment]
The mapbox API key works fine. 

https://leafmap.org/notebooks/38_plotly/
If you run into an error saying "FigureWidget - 'mapbox._derived' Value Error" ([source](https://github.com/plotly/plotly.py/issues/2570#issuecomment-738735816)), uncomment the following line and run it.

```python
leafmap.fix_widget_error()
```

**Comment by krishnaglodha** [comment]
removing `geemap` from requirements.txt is causing https://github.com/opengeos/leafmap/actions/runs/5602386755/jobs/10247560946#step:8:44

**Comment by krishnaglodha** [comment]
Hi @giswqs , you had chance to look at this?

**Comment by giswqs** [comment]
Fixed in #494
